pkg/controller/secretgenerator/secretgenerator_controller.go

```diff
@@ -138,3 +175,80 @@ func (r *ReconcileSecretGenerator) Reconcile(request reconcile.Request) (reconci
 	// Don't requeue if secret is created successfully
 	return reconcile.Result{}, err
 }
+
+// getRoute returns an OpenShift route object. It waits until the .spec.host value exists to avoid possible race conditions, fails otherwise.
+func (r *ReconcileSecretGenerator) getRoute(name string, namespace string) (*routev1.Route, error) {
+	oauthClientRoute := &routev1.Route{}
+	// Get spec.host from route
+	err := wait.PollImmediate(resourceRetryInterval, resourceRetryTimeout, func() (done bool, err error) {
+		err = r.client.Get(context.TODO(), types.NamespacedName{Name: name,
+			Namespace: namespace}, oauthClientRoute)
+		if err != nil {
+			if k8serrors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		} else if oauthClientRoute.Spec.Host == "" {
+			return false, nil
+		} else {
+			return true, nil
+		}
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	return oauthClientRoute, err
+}
+
+func (r *ReconcileSecretGenerator) createOAuthClient(secretName string, name string, secret string, uri string) error {
+	// Create OauthClient resource
+	oauthClient := &ocv1.OAuthClient{
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			Labels: map[string]string{
+				// Add secret label to map OauthClient resource with the secret that triggered its creation.
+				// This is used when deleting the OauthClient resource.
+				secretLabel: secretName,
+			},
+		},
+		Secret:       secret,
+		RedirectURIs: []string{"https://" + uri},
+		GrantMethod:  ocv1.GrantHandlerAuto,
+	}
+
+	err := r.client.Create(context.TODO(), oauthClient)
+	if err != nil {
+		if k8serrors.IsAlreadyExists(err) {
+			log.Infof("Oauth client resource %v already exists", oauthClient.Name)
+			return nil
+		}
+	}
+	return err
+}
+
+func (r *ReconcileSecretGenerator) deleteOAuthClient(secretName string) error {
+	oauthClientList := &ocv1.OAuthClientList{}
+	listopts := []client.ListOption{
+		client.MatchingLabels{secretLabel: secretName}}
```

Comment:
We can change the `createOAuthClient` function to avoid using a LIST operation here. Name the OAuthClient object with the original secret name, instead of using the route name:

```
err = r.createOAuthClient(foundSecret.Name, secret.Value, oauthClientRoute.Spec.Host)

func (r *ReconcileSecretGenerator) createOAuthClient(name string, secret string, uri string) error {
...
}
```

This way, in the `deleteOAuthClient` method you can just retrieve the `OAuthClient` by doing a GET operation (much better performance).